src/lib/testing/expect-saga.ts

```diff
@@ -0,0 +1,192 @@
+
+  private action: Action<Payload> | null = null;
+
+  private finalState: State | null = null;
+
+  private reducer: Reducer<State> | undefined;
+
+  private initialState: undefined | DeepPartial<State> = undefined;
+
+  constructor(saga: Saga<State, Payload>) {
+    this.saga = saga;
+  }
+
+  public withReducer(reducer: Reducer<State>, initialState?: DeepPartial<State>): WithMocksStage<State, Payload> {
+    this.reducer = reducer;
+    this.initialState = initialState;
+
+    return this;
+  }
+
+  public withMocks(mocks: Mocks<State>): AssertionStage<State, Payload> {
+    this.mocks = mocks;
+
+    return this;
+  }
+
+  public toCall<Args extends any[], Return>(
+    func: (...args: Args) => Return,
+    ...args: Args
+  ): AssertionStage<State, Payload> {
+    this.asserts.push({
+      type: 'call',
+      func,
+      args,
+    });
+
+    return this;
+  }
+  public toRun<Args extends any[], Return>(
+    func: SagaFunc<State, Args, Return>,
+    ...args: Args
+  ): AssertionStage<State, Payload> {
+    this.asserts.push({
+      type: 'run',
+      func,
+      args,
+    });
+
+    return this;
+  }
+
+  public toSpawn<Args extends any[], Return>(
+    func: SagaFunc<State, Args, Return>,
+    ...args: Args
+  ): AssertionStage<State, Payload> {
+    this.asserts.push({
+      type: 'spawn',
+      func,
+      args,
+    });
+
+    return this;
+  }
+
+  public toDispatch<DispatchPayload>(action: Action<DispatchPayload>): AssertionStage<State, Payload> {
+    this.asserts.push({
+      type: 'dispatch',
+      action,
+    });
+
+    return this;
+  }
+
+  public toTake<TakePayload>(action: Action<TakePayload>): AssertionStage<State, Payload> {
+    this.asserts.push({
+      type: 'take',
+      action,
+    });
+
+    return this;
+  }
+
+  public dispatch(action: Action<Payload>): FinalStateStage<State, Payload> {
+    this.action = action;
+
+    return this;
+  }
+
+  public toHaveFinalState(state: State): RunStage {
+    this.finalState = state;
+
+    return this;
+  }
+
+  public async run(timeout: number = 10 * 1000) {
+    const { middleware, sagaCompletion } = createTestSagaMiddleware([this.saga], this.asserts, this.mocks);
+
+    const store = createStore(
+      this.reducer || ((state = ({} as unknown) as State) => state),
+      this.initialState,
+      applyMiddleware(middleware),
+    );
+
+    if (this.action === null) {
+      throw new Error('Missing action for starting the saga');
+    }
+
+    store.dispatch(this.action);
+
+    const val = await Promise.race([sagaCompletion(), sleep(timeout)]);
+
+    if (val === 'timeout') {
+      throw new Error(`Saga didn't finish within the timeout of ${timeout / 1000} seconds`);
+    }
+
+    if (this.asserts.length > 0) {
+      throw new Error(`Saga didn't fulfill ${this.asserts.length} assert${this.asserts.length === 1 ? '' : 's'}`);
+    }
+
+    if (this.finalState) {
+      const state = store.getState();
+
+      for (const key in this.finalState) {
```

Comment:
This slipped by me before, probably because we weren't making use of it in the "real" tests so far.

What do you think about having 2 possible functions here e.g. `toHaveFinalState` and `toPartiallyHaveFinalState`. Or as suggested in the other PR support a `map` ("lense") function to "zoom in"?

Reply:
The problem is that this works great if you test your app with your complete reducer which is in most cases a combined reducer. As soon as you test it only with a specific reducer, which works as long as you mock every `select` call, you lose a lot of confidence in your tests.
Especially the developer is most likely not even aware of it while writing the test.

We either need to separate methods for it or maybe an options object, something like `{partial?: boolean}`, though the question would be then whether or not to enable this by default.

Reply:
Let's align this between the two testing functions together in a different PR

Reply:
Ok, we can follow up on this. You're right, testing with a single reducer initially is just setting oneself up for a "shallow" test where more things need to be mocked and it's less realistic.